rfd/0180-sso-mfa.md

```diff
@@ -0,0 +1,925 @@
+default MFA method for UX purposes.
+
+Note: SSO users do not rely on their MFA methods to login, meaning they will
+not be locked out when they have no second factors, even if second factor is
+set to required in the cluster auth preference. SSO users are just required to
+register an MFA device before using any features that require MFA.
+
+Note: Passwordless devices are treated separately from MFA devices. Usually we
+prevent users from deleting passwordless devices to prevent them from being
+locked out, but SSO users without a passwordless device can still perform SSO
+login. Since the SSO MFA device is not a passwordless device, the change above
+will not apply to deleting your last passwordless device. Instead we should
+provide a special case to SSO users to delete their last passwordless device if
+they have any other MFA device. This change is not in direct scope of this RFD,
+but will likely be done beside it.
+
+### Configuration
+
+SSO MFA requires both Cluster Auth Preference and Auth Connector configuration
+to function.
+
+#### Cluster Auth Preference
+
+Whether SSO MFA can be used for a given user depends on whether that user is an
+SSO user and whether that user's Auth Connector is configured for MFA (below).
+Therefore, it does not make sense to add `second_factor: sso` and require SSO
+MFA for every user in the cluster.
+
+| second_factor | otp | webauthn       | sso | mfa mode |
+|---------------|-----|----------------|-----|----------|
+| off           | no  | no             | ?   | off      |
+| optional      | yes | yes (if set)   | ?   | optional |
+| on            | yes | yes (if set)   | ?   | required |
+| otp           | yes | no             | ?   | required |
+| webauthn      | no  | yes (must set) | ?   | required |
+| sso           | ?   | ?              | yes | required |
+
+Instead, we will change the way `second_factor` is configured from a single
+value to multiple permitted `second_factors`. This way, admins have more
+control and less ambiguity over which second factors are permitted.
+
+| second_factors | description                                                         |
+|----------------|---------------------------------------------------------------------|
+| otp            | otp MFA is enabled                                                  |
+| webauthn       | webauthn is enabled, `webauthn` settings must be set                |
+| sso            | sso MFA is enabled for sso-users with an MFA-enabled auth connector |
+
+In order to maintain the ability to set the MFA mode to `off`, `optional`,
+or `required`, we will add a new explicit `mfa_mode` field.
+
+Example config:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  type: oidc
+  connector_name: auth0
+  require_session_mfa: yes
+  mfa_mode: required
+  second_factors: [ sso, webauthn ]
+```
+
+The old `second_factor` field will be deprecated. During the deprecation cycle,
```

Comment:
> The old `second_factor` field will be deprecated

I think we can move pretty slowly here. I also think we can warn about it, but I don't see much point in breaking older configs, however much time passes.